### PR TITLE
librbd: flush pending AIO before resize

### PR DESCRIPTION
**Pull request by dillaman**
Re-add and refactor code for flushing AIO during trim operation.  This prevents in-flight CoR operations from (re-)creating objects in the middle of the resize state machine.

### CODE REVIEW

File: src/librbd/ImageCtx.cc [hunk 1]
```diff
@@ -633,29 +635,46 @@ namespace librbd {
...
     cache_lock.Unlock();
-    int r = flush_cache();
+
+    flush_cache_aio(new FunctionContext(boost::bind(
+      &ImageCtx::invalidate_cache_completion, this, _1, on_finish)));
+  }
+
+  void ImageCtx::invalidate_cache_completion(int r, Context *on_finish) {
     if (r == -EBLACKLISTED) {
       Mutex::Locker l(cache_lock);
```
Comment (jdurgin) [inline src/librbd/ImageCtx.cc]: looks like this ends up being called with cache_lock already held from the ObjectCacher::flush_set() context

Comment (dillaman) [inline src/librbd/ImageCtx.cc]: good catch

File: src/librbd/internal.cc [hunk 2]
```diff
@@ -3549,34 +3554,46 @@ namespace librbd {
...
+      RWLock::RLocker snap_locker(ictx->snap_lock);
+      snap_id = ictx->snap_id;
+      snapc = ictx->snapc;
+
+      // map
+      for (vector<pair<uint64_t,uint64_t> >::const_iterator p =
+             image_extents.begin();
+	   p != image_extents.end(); ++p) {
+        uint64_t len = p->second;
+        ictx->snap_lock.get_read();
```
Comment (jdurgin) [inline src/librbd/internal.cc]: this is now already locked

Comment (jdurgin) [inline src/librbd/internal.cc]: this is now already locked

**Comment by jdurgin** [comment]
This looks good. I noticed the DiffIterateStress test is taking a long time, even as a unit test. Maybe from an earlier change to async scheduling after a delay for some code paths?

**Pull request merged**